frb/dlas.py

```diff
@@ -0,0 +1,250 @@
+""" Module for assessing impact of intervening galaxies
+   (DLAs) on FRB measurements
+   Based on calclations presented in Prochaska & Neeleman 2017
+"""
+from __future__ import print_function, absolute_import, division, unicode_literals
+
+import numpy as np
+import pdb
+
+from scipy.interpolate import interp1d
+
+from astropy import units as u
+
+from .io import load_dla_fits
+
+
+def approx_avgDM(zeval, dla_model='atan', verbose=False):
+    """ Calculate the average DM from intervening galaxies
+    This method is approximate (and fast) and accurate
+    to better than 1% in-so-far as the analysis is correct.
+
+    From Prochaska & Neeleman 2017
+
+    Parameters
+    ----------
+    zeval : float or ndarray
+      Redshift(s) for evaluation
+    dla_model : str, optional
+
+    Returns
+    -------
+    avgDM : Quantity (depending on type of input z)
+      Units of pc/cm**3
+
+    """
+    # Init
+    mlz = _model_lz(dla_model)
+    if isinstance(zeval, float):
+        flg_float = True
+        zeval = np.array([zeval])
+    else:
+        flg_float = False
+    # Error on upper bound
+    if np.max(zeval) > 5.:
+        raise IOError("Calculation is only valid to z=5")
+    # Calculate
+    zcalc = np.linspace(0., 5., 10000)
+    dz = np.median(zcalc-np.roll(zcalc,1))
+    # Load DLA fits model
+    dla_fits = load_dla_fits()
+    # Evaluate l(z)
+    lz = mlz['eval'](zcalc)
+
+    # Average NHI
+    avgNHI = _avgN_dbl_pow(dla_fits=dla_fits)
+
+    # Take ne/nH
+    nenH_p = dla_fits['nenH']['loglog']
+    nenH = nenH_p['bp'] + nenH_p['m'] * (avgNHI - 20.3)
+
+    # Integrate lz for n(z)
+    cumul = np.cumsum(lz * dz)
+
+    # Average <z>
+    avgz = np.cumsum(zcalc * lz * dz) / cumul
+
+    '''
+    # <DM> for a single DLA (rest-frame)
+    DM_DLA = 10. ** (avgNHI + nenH) / u.cm ** 2
+    if verbose:
+        print("DM for an average DLA = {} (rest-frame)".format(DM_DLA.to('pc/cm**3')))
+    '''
+
+    # Altogether now
+    avgDM_values = 10 ** avgNHI * 10 ** nenH * cumul / (1 + avgz) #/ u.cm ** 2
+
+    # Finish up
+    DM_values = np.zeros_like(zeval)
+    for kk,iz in enumerate(zeval):
+        iminz = np.argmin(np.abs(iz - zcalc))
+        DM_values[kk] = avgDM_values[iminz]
+    # Return
+    return (DM_values / u.cm**2).to('pc/cm**3')
+
+
+def monte_DM(zeval, model='atan', nrand=100, verbose=False):
+    """
+    Parameters
+    ----------
+    zeval : ndarray
```

Comment:
I suggest to allow for single values as well. Having to type np.array([2.3]) to get the DM is cumbersome.

Reply:
ok, but the returned DMs will always be a multi-dimensionsal array